### PR TITLE
fix: Set the daily schedule to a specific time, rather than 24hr from start up

### LINKED ISSUE #3454: [BUG] - Recipe timeline event is not created automatically after recipe was added to the meal planner
### First Check

- [X] This is not a feature request.
- [X] I added a very descriptive title to this issue (title field is above this).
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Mealie documentation, with the integrated search.
- [X] I already read the docs and didn't find an answer.
- [X] This issue can be replicated on the demo site (https://demo.mealie.io/).

### What is the issue you are experiencing?

According to https://github.com/mealie-recipes/mealie/pull/2050 a recipe timeline event will be created automatically on a daily basis after a recipe was added to the meal planner. 

On my instance that never happened, both the 'last made' date and the recipe timeline doesn't change.
I found this older related issue https://github.com/mealie-recipes/mealie/issues/2344  

I checked with v1.4.0, the latest nightly and the demo site. 
Maybe I'm missing something? Thanks for your help 🙏

### Steps to Reproduce

- Add a recipe to the meal planner as mentioned in https://github.com/mealie-recipes/mealie/issues/2344
> Added recipe for today from Meal Plan page
> Added recipe for today to Meal Plan from the Recipe
> Added recipe for today to Meal Plan from the recipe overview

- Wait for at least one day or longer.

- The 'last made' date of the recipe does not change, no recipe timeline event was added.

### Please provide relevant logs

I could not find anything related to this issue or any errors.
The only scheduled task I can see the logs is the hourly reset of locked users, but no hint of timeline events.
`INFO: 11-Apr-24 23:02:51 	locked users reset`

### Mealie Version

Version
v1.4.0
Build
[f709d11952ca8552214e4b9e8c2a189bc11bfb2a](https://github.com/mealie-recipes/mealie/commit/f709d11952ca8552214e4b9e8c2a189bc11bfb2a)

Version
nightly
Build
[94678fe6e0a4831b3cbc24b366dba1009d360f23](https://github.com/mealie-recipes/mealie/commit/94678fe6e0a4831b3cbc24b366dba1009d360f23)

### Deployment

Docker (Linux)

### Additional Deployment Details

_No response_

### PR DESCRIPTION
**Pull request by boc-the-git**
## What type of PR is this?

- bug

## What this PR does / why we need it:

The previous logic for running daily tasks was that at container start it would NOT run, but sleep for 24hrs and then run. For people (like me) who might restart their container daily for backups, or who regularly update their image, this means they effectively never get the daily tasks run.
This change makes it so the daily tasks are scheduled to run at the next 23:45 (or other configured time). Meaning if a container is started at 22:00, it'll run 1:45 later (as opposed to 24:00 later).

This uses changes by @Grygon on https://github.com/mealie-recipes/mealie/pull/2497 as its basis.

## Which issue(s) this PR fixes:

Maybe fixes #3454; it's the reason I noticed this.

## Special notes for your reviewer:

Some comments will be added against the code.

## Testing

Have done a bit on my prod system.

Realistic case:
```
mealie  | DEBUG    2024-05-24T20:37:46 - Current time is 2024-05-24 20:37:46.042663 and DAILY_SCHEDULE_TIME is 23:45
mealie  | DEBUG    2024-05-24T20:37:46 - Hours until 3 and minutes until 8
mealie  | INFO     2024-05-24T20:37:46 - Daily tasks scheduled for 2024-05-24 23:45:00
```

Test case, showing it rolls to next day as needed:
```
mealie  | DEBUG    2024-05-24T19:47:06 - Current time is 2024-05-24 19:47:06.369710 and target time is 18:45
mealie  | DEBUG    2024-05-24T19:47:06 - Hours until 23 and minutes until 58
mealie  | INFO     2024-05-24T19:47:06 - Daily tasks scheduled for 2024-05-25 19:45:00
```

Multi day Production run (note changes were implemented at 2024-05-24T20:37:46). This shows that the daily tasks are being run every 23:45 as expected despite, importantly, my container restarting at 1am each day.
```
mealie  | INFO     2024-05-24T19:30:07 - Uvicorn running on http://0.0.0.0:9000 (Press CTRL+C to quit)
mealie  | INFO     2024-05-24T19:45:32 - Uvicorn running on http://0.0.0.0:9000 (Press CTRL+C to quit)
mealie  | INFO     2024-05-24T19:47:06 - Uvicorn running on http://0.0.0.0:9000 (Press CTRL+C to quit)
mealie  | INFO     2024-05-24T20:07:01 - Uvicorn running on http://0.0.0.0:9000 (Press CTRL+C to quit)
mealie  | INFO     2024-05-24T20:17:42 - Uvicorn running on http://0.0.0.0:9000 (Press CTRL+C to quit)
mealie  | INFO     2024-05-24T20:37:46 - Uvicorn running on http://0.0.0.0:9000 (Press CTRL+C to quit)
mealie  | DEBUG    2024-05-24T23:45:00 - Running daily callbacks
mealie  | INFO     2024-05-25T01:00:14 - Uvicorn running on http://0.0.0.0:9000 (Press CTRL+C to quit)
mealie  | DEBUG    2024-05-25T23:45:00 - Running daily callbacks
mealie  | INFO     2024-05-26T01:00:15 - Uvicorn running on http://0.0.0.0:9000 (Press CTRL+C to quit)
```

### CODE REVIEW

File: mealie/services/scheduler/scheduler_service.py [hunk 1]
```diff
@@ -18,18 +21,43 @@ class SchedulerService:
...
+    await asyncio.sleep(wait_seconds)
+    await run_daily()
+
 
 def _scheduled_task_wrapper(callable):
     try:
         callable()
     except Exception as e:
-        logger.error(f"Error in scheduled task func='{callable.__name__}': exception='{e}'")
+        logger.error("Error in scheduled task func='%s': exception='%s'", callable.__name__, e)
```
Comment (boc-the-git) [inline mealie/services/scheduler/scheduler_service.py:67]: Linter

File: mealie/services/scheduler/scheduler_service.py [hunk 2]
```diff
@@ -18,18 +21,43 @@ class SchedulerService:
...
 
 
-@repeat_every(minutes=MINUTES_DAY, wait_first=True, logger=logger)
+@repeat_every(minutes=MINUTES_DAY, wait_first=False, logger=logger)
```
Comment (boc-the-git) [inline mealie/services/scheduler/scheduler_service.py:70]: We already handle the schedule elsewhere, so we don't want to wait on the first call to run_daily

File: mealie/services/scheduler/scheduler_service.py [hunk 3]
```diff
@@ -18,18 +21,43 @@ class SchedulerService:
...
+
+    hours_until = ((hour_target - now.hour) % 24) or 24
+    minutes_until = (minute_target - now.minute) % 60
+    logger.debug("Hours until %s and minutes until %s", str(hours_until), str(minutes_until))
+
+    delta = timedelta(hours=hours_until, minutes=minutes_until)
+    target_time = (now + delta).replace(microsecond=0, second=0)
+    logger.info("Daily tasks scheduled for %s", str(target_time))
+    wait_seconds = (target_time - now).total_seconds()
+    await asyncio.sleep(wait_seconds)
```
Comment (boc-the-git) [inline mealie/services/scheduler/scheduler_service.py:53]: Here we sleep to the first scheduled time, which is why we don't want to "wait" on the first run_daily - this is where that happens.

File: mealie/services/scheduler/scheduler_service.py [hunk 4]
```diff
@@ -18,18 +21,43 @@ class SchedulerService:
...
+
+async def schedule_daily():
+    now = datetime.now()
+    daily_schedule_time = get_app_settings().DAILY_SCHEDULE_TIME
+    logger.debug(
+        "Current time is %s and DAILY_SCHEDULE_TIME is %s",
+        str(now),
+        daily_schedule_time,
+    )
+    hour_target = int(daily_schedule_time.split(":")[0])
```
Comment (michael-genson) [inline mealie/services/scheduler/scheduler_service.py]: Can we add a try/catch on the `daily_schedule_time` parsing? Maybe bump it to its own method? That way this doesn't fail if the string is malformed. We could probably default to midnight and log the exception:

```python
try:
    hour_target, minute_target = _parse_daily_schedule_time(daily_schedule_time)
except Exception as e:
    logger.exception(f"Unable to parse {daily_schedule_time=})
    hour_target = minute_target = 0
```

File: docs/docs/documentation/getting-started/installation/backend-config.md [hunk 5]
```diff
@@ -16,7 +16,8 @@
 | TZ                            |          UTC          | Must be set to get correct date/time on the server                                  |
 | ALLOW_SIGNUP<super>\*</super> |         false         | Allow user sign-up without token                                                    |
 | LOG_CONFIG_OVERRIDE           |                       | Override the config for logging with a custom path                                  |
-| LOG_LEVEL                     |         info          | logging level configured                                                            |
+| LOG_LEVEL                     |         info          | Logging level configured                                                            |
+| DAILY_SCHEDULE_TIME           |        23:45          | The time of day to run the daily tasks.                                |
```
Comment (michael-genson) [inline docs/docs/documentation/getting-started/installation/backend-config.md]: Can we extend the last pipe out so it lines up? Just for my own sanity, I don't think it actually matters

**Review by michael-genson — CHANGES_REQUESTED** [review]
Looks good! Two small changes

File: docs/docs/documentation/getting-started/installation/backend-config.md (hunk 5 above)
Comment (boc-the-git) [reply docs/docs/documentation/getting-started/installation/backend-config.md]: I'd rather not try to be pedantic about this .. that whole table structure is a bit of a mess, and any change made to try and fix it now will just get "broken" in future. There's a ton of inconsistency in there already.

Comment (michael-genson) [reply docs/docs/documentation/getting-started/installation/backend-config.md]: Huh, maybe it's just GitHub then, it seems to be formatted correctly to me:

image

But whatever it really doesn't matter anyway

Comment (boc-the-git) [reply docs/docs/documentation/getting-started/installation/backend-config.md]: No, you're correct. VSCode was completely distorting it for me 🤦 

Fixed.

File: mealie/services/scheduler/scheduler_service.py (hunk 4 above)
Comment (boc-the-git) [reply mealie/services/scheduler/scheduler_service.py]: Updated. Test:

```
mealie  | DEBUG    2024-05-26T13:20:41 - Current time is 2024-05-26 13:20:41.016770 and DAILY_SCHEDULE_TIME is ThisIsBadData
mealie  | ERROR    2024-05-26T13:20:41 - Unable to parse daily_schedule_time='ThisIsBadData'
mealie  | Traceback (most recent call last):
mealie  |   File "/app/mealie/services/scheduler/scheduler_service.py", line 39, in schedule_daily
mealie  |     hour_target, minute_target = _parse_daily_schedule_time(daily_schedule_time)
mealie  |   File "/app/mealie/services/scheduler/scheduler_service.py", line 60, in _parse_daily_schedule_time
mealie  |     hour_target = int(time.split(":")[0])
mealie  | ValueError: invalid literal for int() with base 10: 'ThisIsBadData'
mealie  | ERROR    2024-05-26T13:20:41 - Unable to parse daily_schedule_time='ThisIsBadData'
mealie  | Traceback (most recent call last):
mealie  |   File "/app/mealie/services/scheduler/scheduler_service.py", line 39, in schedule_daily
mealie  |     hour_target, minute_target = _parse_daily_schedule_time(daily_schedule_time)
mealie  |   File "/app/mealie/services/scheduler/scheduler_service.py", line 60, in _parse_daily_schedule_time
mealie  |     hour_target = int(time.split(":")[0])
mealie  | ValueError: invalid literal for int() with base 10: 'ThisIsBadData'
mealie  | DEBUG    2024-05-26T13:20:41 - Hours until 10 and minutes until 25
mealie  | INFO     2024-05-26T13:20:41 - Daily tasks scheduled for 2024-05-26 23:45:00
mealie  | INFO     2024-05-26T13:20:41 - Application startup complete.
mealie  | INFO     2024-05-26T13:20:41 - Uvicorn running on http://0.0.0.0:9000 (Press CTRL+C to quit)
```

**Review by michael-genson — APPROVED** [review]
LGTM!

**Pull request merged**